src/hatch_fancy_pypi_readme/_config.py

```diff
@@ -19,36 +22,97 @@ class Config:
     substitutions: list[Substituter]
 
 
+SCHEMA = {
+    "type": "object",
+    "properties": {
+        "content-type": {
+            "type": "string",
+            "enum": ["text/markdown", "text/x-rst"],
+        },
+        "fragments": {
+            "type": "array",
+            "minItems": 1,
+            # Items are validated separately for better error messages.
+            "items": {"type": "object"},
+        },
+        "substitutions": {
+            "type": "array",
+            "items": {
+                "type": "object",
+                "properties": {
+                    "pattern": {"type": "string", "format": "regex"},
```

Comment:
To speak out what I was saying whenever -- newer drafts have stuff called vocabularies, which are basically ways to package up bundles of extension keywords. There's only partial support for doing that in my lib so far, but what's there is enough for this kind of thing if you ever need it. Specifically:

```
import jsonschema
import re

def is_regex(validator, regex, instance, schema):
    if not validator.is_type(instance, "string"):
        return
    try:
        re.compile(instance)
    except (re.error, ValueError):
        if regex:
            yield jsonschema.ValidationError(f"{instance} is not a Python regular expression")
    else:
        if not regex:
            yield jsonschema.ValidationError(f"{instance} is a Python regular expression")

WithRegexValidator = jsonschema.validators.extend(jsonschema.validators.Draft202012Validator, validators={"regex": is_regex})
```

means you can write `{"type": "string", "regex": True}` in your schemas say, so:

```
v = WithRegexValidator({"type": "string", "regex": True})
v.validate("a;.$][sdf")
```

produces:

```
Traceback (most recent call last):
  File "/Users/julian/Development/hatch-fancy-pypi-readme/foo.py", line 18, in <module>
    v.validate("a;.$][sdf")
  File "/Users/julian/.dotfiles/.local/share/virtualenvs/jsonschema/lib/python3.10/site-packages/jsonschema/validators.py", line 303, in validate
    raise error
jsonschema.exceptions.ValidationError: a;.$][sdf is not a Python regular expression

Failed validating 'regex' in schema:
    {'regex': True, 'type': 'string'}

On instance:
    'a;.$][sdf'
```

Reply:
what is `regex` in `is_regex` and why is `not regex` a regular expression? 😅

and what's the upside over `"format": "regex"` under-specification aside? is the goal to ideally not use format at all?

Reply:
ah it's the argument…so the key name determines both the validator and its value is passed into the validator as second posarg?

Reply:
> ah it's the argument…so the key name determines both the validator and its value is passed into the validator as second posarg?

Yes (the name doesn't matter, today it's passed positionally, call it whatever)

> and what's the upside over "format": "regex" under-specification aside? is the goal to ideally not use format at all?

Basically this exact nonsense about ignoring unknown formats. The spec mandates implementations (other implementations) ignore your `regex` format if it doesn't know what to do with it, but you can specify your `regex` keyword as you wish, and there's a mechanism to tell implementations to hard fail if they don't understand what that keyword means from within your schema (which I didn't show, that's the bit that's related to vocabularies)

But functionally for my implementation obviously there's no difference as an end user.

Reply:
OK I got it working, although I had to remove the `if not validator.is_type(instance, "string"):` part because i just could't get it to trigger…apparently the string test comes first?